swim/heal_partition.go

```diff
@@ -0,0 +1,147 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import "time"
+
+// AttemptHeal attempts to heal a partition between the node and the target.
+//
+// Be mindfull that calling this function will not result in a heal when there
+// are nodes that need to be reincarated to take precedence over the faulty
+// declarations that occur during a network partition. A cluster may therefore
+// need multiple calls to this function with some time in between to heal.
+//
+// Check out ringpop-common/docs for a full description of the algorithm.
+func AttemptHeal(node *Node, target string) ([]string, error) {
+	node.emit(AttemptHealEvent{})
+	node.logger.WithField("target", target).Info("attempt heal")
+
+	// If join request succeeds a partition is detected,
+	// this node will now coordinate the healing mechanism.
+	joinRes, err := sendJoinRequest(node, target, time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	A := node.disseminator.MembershipAsChanges()
+	B := joinRes.Membership
+
+	// Get the nodes that aren't mergeable and need to reincarnate
+	csForA, csForB := nodesThatNeedToReincarnate(A, B)
+
+	// Reincarnate the nodes that need to reincarnate
+	if len(csForA) != 0 || len(csForB) != 0 {
+		err = reincarnateNodes(node, target, csForA, csForB)
+		return aliveHosts(B), err
+	}
+
+	// Merge partitions if no node needs to reincarnate
+	err = mergePartitions(node, target, B)
+	return aliveHosts(B), err
+}
+
+// nodesThatNeedToReincarnate finds all nodes would become faulty when
+// membership A gets merged with membership B. These need to reincarnate
+// first, before we can heal the partition with a merge.
+func nodesThatNeedToReincarnate(A, B []Change) (csForA, csForB []Change) {
+	// Find changes that are alive for B and faulty for A and visa versa.
+	for _, b := range B {
+		a, ok := selectMember(A, b.Address)
+		if !ok {
+			continue
+		}
+
+		// faulty for partition A, alive for partition B
+		// needs reincarnation in partition B.
+		if a.Status == Faulty && b.Status == Alive && a.Incarnation >= b.Incarnation {
```

Comment:
This breaks in the case where it is a suspect or a tombstone that could still contaminate the other partition. Can we use state precedence here in combination with the incarnation number to figure out what to do?